cmd/broker/ingress/main.go

```diff
@@ -70,26 +75,46 @@ func main() {
 	}
 
 	err = mgr.Add(&runnableServer{
-		logger: logger,
-		s:      s,
+		logger:          logger,
+		s:               s,
+		ShutdownTimeout: writeTimeout,
 	})
 	if err != nil {
 		logger.Fatal("Unable to add runnableServer", zap.Error(err))
 	}
 
+	// Metrics
+	e, err := prometheus.NewExporter(prometheus.Options{Namespace: metricsNamespace})
+	if err != nil {
+		logger.Fatal("Unable to create Prometheus exporter", zap.Error(err))
+	}
+	view.RegisterExporter(e)
+	sm := http.NewServeMux()
+	sm.Handle("/metrics", e)
+	metricsSrv := &http.Server{
+		Addr:         fmt.Sprintf(":%d", metricsPort),
+		Handler:      e,
+		ErrorLog:     zap.NewStdLog(logger),
+		ReadTimeout:  readTimeout,
+		WriteTimeout: writeTimeout,
+	}
+
+	err = mgr.Add(&runnableServer{
+		s:               metricsSrv,
+		logger:          logger,
+		ShutdownTimeout: writeTimeout,
+	})
+	if err != nil {
+		logger.Fatal("Unable to add metrics runnableServer", zap.Error(err))
+	}
+
 	// Set up signals so we handle the first shutdown signal gracefully.
 	stopCh := signals.SetupSignalHandler()
 	// Start blocks forever.
 	if err = mgr.Start(stopCh); err != nil {
 		logger.Error("manager.Start() returned an error", zap.Error(err))
 	}
 	logger.Info("Exiting...")
```

Comment:
I think Go programs exit as soon as main() finishes. As we no longer wait on Shutdown() inside main(), will this forcibly exit, without waiting for shutdown to do so gracefully?

Reply:
It will forcibly exit. I refactored this and added a WaitGroup to avoid this. The real fix is to change the controller-runtime Manager to wait for all Runnables to stop before returning from Start, as described in https://github.com/kubernetes-sigs/controller-runtime/issues/350.